File: backend/app/protokolle/pruefliste/parameter.py
"""What a search box and a page number mean, as values.

No SQLAlchemy and no database in here. The review queue's query is built from
these answers, and every one of them has a wrong answer available to it: a search
that matches everything, a page of nothing, a pager that says "Seite 1 von 0".
Keeping them as plain functions is what lets each be held to its promise without
a database, which is the arrangement coding-standards.md asks for.
"""

from datetime import date

#: What LIKE reads as "one character" and "any number of characters".
_PLATZHALTER = ("%", "_")

#: The character that takes a wildcard's meaning away, passed to ILIKE as its
#: ESCAPE. Doubled inside a search term for the reason maskiere_platzhalter
#: gives.
ESCAPE = "\\"

#: How many rows a request asks for when it does not say.
PRO_SEITE_STANDARD = 25

#: One row. A page of nothing would be a pager with no last page.
PRO_SEITE_MIN = 1

#: The most one request may ask for. The queue is read by a person a screenful at
#: a time, and the cap is what stops one request asking for every protocol FFS
#: has ever taken.
PRO_SEITE_MAX = 100

#: The years date() can express. Not a statement about how old a Befischung may
#: be: it is only what keeps jahresgrenzen from raising on a year nobody meant.
JAHR_MIN = date.min.year
JAHR_MAX = date.max.year


def maskiere_platzhalter(text: str) -> str:
    """One search term with LIKE's wildcards taken out of it.

    Without this, typing 50% into the search box matches every protocol in the
    database, because % means "anything". The surveyor sees a search box that
    ignores what they type.

    The escape character goes first, and the order is the whole point: masking %
    before \\ would escape the backslash this function had just added and turn
    \\% back into a wildcard.
    """
    maskiert = text.replace(ESCAPE, ESCAPE * 2)
    for zeichen in _PLATZHALTER:
        maskiert = maskiert.replace(zeichen, ESCAPE + zeichen)
    return maskiert


def suchbegriffe(suche: str | None) -> tuple[str, ...]:
    """A search box split into the words that each have to find something.

    Every term is required, and each may match any of the searched columns. That
    is what makes "Schussen Weissenau" find the row whose water is Schussen and
    whose Ortsangabe is Weissenau. Matching the typed string as one value would
    find nothing, since no single column holds both words, and the box would look
    broken to the one person most likely to try it.

    Whitespace only is the same as no search at all, so a box somebody tabbed
    through does not narrow the list to nothing.
    """
    if suche is None:
        return ()
    return tuple(suche.split())


def suchmuster(begriff: str) -> str:
    """One term as ILIKE is asked for it: anywhere inside the value.

    The surrounding % are the wildcards we mean. Anything the person typed is
    masked first, so theirs are not.
    """
    return f"%{maskiere_platzhalter(begriff)}%"


def begrenze_seite(seite: int) -> int:
    """A page number that exists. Pages are 1-based and there is none before."""
    return max(seite, 1)


def begrenze_pro_seite(pro_seite: int) -> int:
    """A page size between one row and the cap."""
    return min(max(pro_seite, PRO_SEITE_MIN), PRO_SEITE_MAX)


def versatz(seite: int, pro_seite: int) -> int:
    """How many rows to skip to reach this page."""
    return (begrenze_seite(seite) - 1) * begrenze_pro_seite(pro_seite)


def seitenzahl(gesamt: int, pro_seite: int) -> int:
    """How many pages a total makes, never fewer than one.

    An empty queue has one empty page, so the pager reads "Seite 1 von 1". Zero
    pages would be a count no page number could satisfy, and the screen would
    have to special-case it.
    """
    if pro_seite < PRO_SEITE_MIN:
        return 1
    return max(-(-gesamt // pro_seite), 1)


def jahresgrenzen(jahr: int) -> tuple[date, date]:
    """A year as the first and the last day of it.

    Two dates rather than an extract(year) on the column, because a comparison
    against a plain date column can use an index and a function over that column
    cannot. There is no index on datum today, and this is the shape that does not
    have to change on the day there is one.
    """
    return date(jahr, 1, 1), date(jahr, 12, 31)

File: frontend/src/protokoll/pruefliste/parameter.ts
/* What the address bar says the review queue is.
 *
 * Plain values in and out: no React, no fetch, no i18next. The screen reads a
 * URLSearchParams through abfrageAus, hands one back through alsSuchparameter,
 * and the request is built by alsEndpunktParameter. Every one of them has a
 * wrong answer available to it, which is what coding-standards.md asks be kept
 * testable without a browser.
 *
 * **The address bar is this screen's state, and that is load-bearing.** Holding
 * the filters in a component instead would work today and break feature 12d:
 * Vorheriges and Naechstes have to rebuild this exact list from a protocol's URL
 * alone, and a filter living in a component would make "the protocol before this
 * one" depend on what the reviewer happened to click earlier. It also means a
 * reload comes back to the same list and a filtered queue can be sent as a link.
 *
 * A value out of the address bar is untrusted input, the same rule auth/weiter.ts
 * already states and tests. Anything unrecognised becomes the default rather than
 * being passed on: sent to the endpoint it would be answered with a 422, and a
 * hand-edited address would look like a broken page.
 */

/* The six states the queue may be asked for, exactly as Pruefstatus in
   backend/app/api/schemas.py spells them. DRAFT is not among them: the queue
   lists work that has been handed in, and the endpoint refuses to be asked. */
export const PRUEFSTATUS = [
  'SUBMITTED',
  'IN_REVIEW',
  'NEEDS_CHANGES',
  'REJECTED',
  'ACCEPTED',
  'LOCKED',
] as const

export type Pruefstatus = (typeof PRUEFSTATUS)[number]

/* The protocols nobody has decided on yet, which is what a work queue is for.
 *
 * The endpoint takes status repeatably precisely so this screen can ask for both
 * at once. It is offered as a named entry in the dropdown rather than applied
 * quietly, because a queue that hides rows without saying so reads as a database
 * with protocols missing from it. */
export const OFFEN = ['SUBMITTED', 'IN_REVIEW'] as const

/** What the Status dropdown offers: the open pair, everything, or one state. */
export type Statuswahl = 'offen' | 'alle' | Pruefstatus

/* The four orders Sortierung in backend/app/protokolle/pruefliste/dienst.py
   declares, and no others. A fifth here would be a sort the server has not got,
   which is a control that silently does nothing. */
export const SORTIERUNGEN = [
  'eingereicht_alt',
  'eingereicht_neu',
  'datum_neu',
  'gewaesser',
] as const

export type Sortierung = (typeof SORTIERUNGEN)[number]

/* Mirrors SEITE_MAX in backend/app/protokolle/pruefliste/parameter.py. Clamped
   here as well as there so a silly number in the address bar never becomes a
   request at all. */
const SEITE_MAX = 1_000_000

/* The years a date can express, as the backend's JAHR_MIN and JAHR_MAX have
   them. Not a statement about how old a Befischung may be: it is what keeps a
   nonsense year out of a request the endpoint would refuse. Which years the
   dropdown offers is a separate and narrower question. */
const JAHR_MIN = 1
const JAHR_MAX = 9999

/** The whole of what this screen is showing, as values. */
export interface Prueflistenabfrage {
  status: Statuswahl
  anlass: string | null
  jahr: number | null
  /** As typed, untrimmed. Trimmed on the way to the endpoint, not on the way in. */
  suche: string
  /** One species export code, as the catch table stores it, e.g. "HECH". */
  art: string | null
  sortierung: Sortierung
  seite: number
}

/** Nothing narrowed, longest wait first, page one. */
export const STANDARD: Prueflistenabfrage = {
  status: 'offen',
  anlass: null,
  jahr: null,
  suche: '',
  art: null,
  sortierung: 'eingereicht_alt',
  seite: 1,
}

function begrenzeSeite(seite: number): number {
  if (!Number.isFinite(seite)) return 1
  return Math.min(Math.max(Math.floor(seite), 1), SEITE_MAX)
}

function statusAus(roh: string | null): Statuswahl {
  if (roh === 'alle') return 'alle'
  return (PRUEFSTATUS as readonly string[]).includes(roh ?? '')
    ? (roh as Pruefstatus)
    : STANDARD.status
}

function sortierungAus(roh: string | null): Sortierung {
  return (SORTIERUNGEN as readonly string[]).includes(roh ?? '')
    ? (roh as Sortierung)
    : STANDARD.sortierung
}

function jahrAus(roh: string | null): number | null {
  if (roh === null) return null

  const jahr = Number(roh)
  if (!Number.isInteger(jahr) || jahr < JAHR_MIN || jahr > JAHR_MAX) return null
  return jahr
}

function seiteAus(roh: string | null): number {
  if (roh === null) return STANDARD.seite

  const seite = Number(roh)
  return Number.isNaN(seite) ? STANDARD.seite : begrenzeSeite(seite)
}

/* The species, or nothing.
 *
 * The code is not checked against the form's list. The picker can only offer
 * real ones, and a list belongs to a form version: a protocol frozen on an older
 * one may legitimately name a code today's list no longer carries, and the
 * endpoint answers an unknown code with an empty page rather than a refusal.
 * Whitespace alone is nobody's choice and becomes no filter, the same as an empty
 * search box.
 */
function artAus(roh: string | null): string | null {
  return roh === null || roh.trim() === '' ? null : roh.trim()
}

function sucheAus(roh: string | null): string {
  // Whitespace alone is the same as no search, so a box somebody tabbed through
  // does not narrow the queue to nothing.
  return roh === null || roh.trim() === '' ? '' : roh
}

/** The address bar, read as a selection. */
export function abfrageAus(parameter: URLSearchParams): Prueflistenabfrage {
  return {
    status: statusAus(parameter.get('status')),
    anlass: parameter.get('anlass'),
    jahr: jahrAus(parameter.get('jahr')),
    suche: sucheAus(parameter.get('suche')),
    art: artAus(parameter.get('art')),
    sortierung: sortierungAus(parameter.get('sortierung')),
    seite: seiteAus(parameter.get('seite')),
  }
}

/* The selection, as the address bar should carry it.
 *
 * Everything at its default is left out, so the plain queue has no query string
 * at all and a shared link says only what was actually narrowed.
 */
export function alsSuchparameter(abfrage: Prueflistenabfrage): URLSearchParams {
  const parameter = new URLSearchParams()

  if (abfrage.status !== STANDARD.status) parameter.set('status', abfrage.status)
  if (abfrage.anlass !== null) parameter.set('anlass', abfrage.anlass)
  if (abfrage.jahr !== null) parameter.set('jahr', String(abfrage.jahr))
  if (abfrage.suche !== '') parameter.set('suche', abfrage.suche)
  if (abfrage.art !== null) parameter.set('art', abfrage.art)
  if (abfrage.sortierung !== STANDARD.sortierung) {
    parameter.set('sortierung', abfrage.sortierung)
  }
  if (abfrage.seite !== STANDARD.seite) parameter.set('seite', String(abfrage.seite))

  return parameter
}

/** One or more parts of the selection, changed together. */
export type Aenderung = Partial<Prueflistenabfrage>

/** Everything that narrows or reorders the list, as opposed to paging it. */
const FILTERFELDER = ['status', 'anlass', 'jahr', 'suche', 'art', 'sortierung'] as const

/* Whether the reader has narrowed the queue themselves.
 *
 * The page number is deliberately not part of the answer, and that is the whole
 * reason this is a named, tested function rather than a check on the query
 * string. Asking "is anything in the address bar?" counts being on page 2 as
 * being filtered, which makes an empty page past the end say "nothing matches
 * your filters" to somebody who has set none, and offer them a reset button for
 * filters that do not exist.
 *
 * The order is counted, because a reader who changed it has made a choice about
 * the list that is worth offering to undo alongside the rest.
 */
export function istGefiltert(abfrage: Prueflistenabfrage): boolean {
  return FILTERFELDER.some((feld) => abfrage[feld] !== STANDARD[feld])
}

/* A changed selection, back on the first page whenever the list itself changed.
 *
 * Without the reset, narrowing the queue while on page 4 lands on a page that no
 * longer exists, and a filter that matched plenty looks like one that matched
 * nothing. Setting a filter to the value it already holds is not a change, so
 * reselecting the order you are already using does not throw away your place.
 */
export function mitAenderung(
  abfrage: Prueflistenabfrage,
  aenderung: Aenderung,
): Prueflistenabfrage {
  const neu = { ...abfrage, ...aenderung }
  const gefiltert = FILTERFELDER.some((feld) => neu[feld] !== abfrage[feld])

  return { ...neu, seite: begrenzeSeite(gefiltert ? STANDARD.seite : neu.seite) }
}

/* The selection, as GET /api/v1/pruefliste is asked for it.
 *
 * Not the same string as the address bar's. The order and the page are always
 * named, defaults included, because nobody reads this one and a request in the
 * network tab that spells out what it asked for answers for itself. The status
 * is expanded here: "offen" is two repeated parameters, "alle" is none at all.
 */
export function alsEndpunktParameter(abfrage: Prueflistenabfrage): URLSearchParams {
  const parameter = new URLSearchParams()

  if (abfrage.status === 'offen') {
    for (const zustand of OFFEN) parameter.append('status', zustand)
  } else if (abfrage.status !== 'alle') {
    parameter.append('status', abfrage.status)
  }

  if (abfrage.anlass !== null) parameter.set('anlass', abfrage.anlass)
  if (abfrage.jahr !== null) parameter.set('jahr', String(abfrage.jahr))

  const suche = abfrage.suche.trim()
  if (suche !== '') parameter.set('suche', suche)

  if (abfrage.art !== null) parameter.set('art', abfrage.art)

  parameter.set('sortierung', abfrage.sortierung)
  parameter.set('seite', String(abfrage.seite))

  return parameter
}
